component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java

```diff
@@ -336,6 +335,45 @@ protected boolean authenticate(X509Certificate registeredCert, X509Certificate r
         return trustedCert;
     }
 
+    /**
+     * Authenticate the client by comparing the public key of the registered public certificate against the public
+     * key of the certificate presented at TLS hand shake for authentication.
+     *
+     * @param registeredCert X.509 certificate registered at service provider configuration.
+     * @param requestCert    X.509 certificate presented to server during TLS hand shake.
+     * @return Whether the client was successfully authenticated or not.
+     */
+    protected boolean authenticate(X509Certificate registeredCert, X509Certificate requestCert, OAuthAppDO oAuthAppDO)
+            throws OAuthClientAuthnException {
+
+        boolean trustedCert = false;
+        try {
+            String publicKeyOfRegisteredCert = MutualTLSUtil.getThumbPrint(registeredCert, null);
+            String publicKeyOfRequestCert = MutualTLSUtil.getThumbPrint(requestCert, null);
+            if (StringUtils.equals(publicKeyOfRegisteredCert, publicKeyOfRequestCert)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Client certificate thumbprint matched with the registered certificate thumbprint.");
+                }
+                if (StringUtils.isNotEmpty(oAuthAppDO.getTlsClientAuthSubjectDN()) &&
+                        !oAuthAppDO.getTlsClientAuthSubjectDN().equals(requestCert.getSubjectDN().toString())) {
+                    log.debug("Client certificate subjectDN does not match with the registered certificate " +
+                            "subjectDN.");
+                    return false;
+                }
+                trustedCert = true;
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("Client Authentication failed. Client certificate thumbprint did not match with the " +
+                            "registered certificate thumbprint.");
```

Comment:
Lets add the configured cert thumbprint and the received cert thumbprint

Reply:
changed